### PR TITLE
Move modules lists to OSSL_LIB_CTX

### LINKED ISSUE #18226: OSSL_LIB_CTX_load_config is not thread safe
When OSSL_LIB_CTX_load_config is called simultaneously from multiple threads, the program abort with the following trace:

 ```
(gdb) where
#0  0x00007ffff7623387 in raise () from /lib64/libc.so.6
#1  0x00007ffff7624a78 in abort () from /lib64/libc.so.6
#2  0x00007ffff7665ed7 in __libc_message () from /lib64/libc.so.6
#3  0x00007ffff766c3e4 in malloc_printerr () from /lib64/libc.so.6
#4  0x00007ffff76716e0 in _int_realloc () from /lib64/libc.so.6
#5  0x00007ffff7672d82 in realloc () from /lib64/libc.so.6
#6  0x00000000004fa265 in CRYPTO_realloc (str=0x7ffff0006390, num=104, file=0x734de8 "crypto/stack/stack.c", line=210) at crypto/mem.c:214
#7  0x000000000052e1b8 in sk_reserve (st=0x7ffff0009570, n=1, exact=0) at crypto/stack/stack.c:210
#8  0x000000000052e2f4 in OPENSSL_sk_insert (st=0x7ffff0009570, data=0x7fffe8005ba0, loc=9) at crypto/stack/stack.c:254
#9  0x000000000052e79b in OPENSSL_sk_push (st=0x7ffff0009570, data=0x7fffe8005ba0) at crypto/stack/stack.c:368
#10 0x0000000000496625 in sk_CONF_IMODULE_push (sk=0x7ffff0009570, ptr=0x7fffe8005ba0) at crypto/conf/conf_mod.c:27
#11 0x00000000004970b6 in module_init (pmod=0x7ffff0002c10, name=0x7fffe80037e0 "providers", value=0x7fffe8003820 "provider_sect", cnf=0x7fffe8009770) at crypto/conf/conf_mod.c:390
#12 0x0000000000496b64 in module_run (cnf=0x7fffe8009770, name=0x7fffe80037e0 "providers", value=0x7fffe8003820 "provider_sect", flags=0) at crypto/conf/conf_mod.c:239
#13 0x0000000000496837 in CONF_modules_load (cnf=0x7fffe8009770, appname=0x0, flags=0) at crypto/conf/conf_mod.c:138
#14 0x00000000004969b0 in CONF_modules_load_file_ex (libctx=0x7fffe8009d00, filename=0x70ee50 "/tmp/openssl.cnf", appname=0x0, flags=0)
    at crypto/conf/conf_mod.c:181
#15 0x00000000004f52f8 in OSSL_LIB_CTX_load_config (ctx=0x7fffe8009d00, config_file=0x70ee50 "/tmp/openssl.cnf") at crypto/context.c:234
```
Each thread allocates its own library context and load config.
Please see the sample code below which reproduces the issue.  With mutex held while OSSL_LIB_CTX_load_config() is called, we don't see the issue.

```
#include <stdio.h>
#include <stdlib.h>
#include <openssl/bio.h>
#include <openssl/err.h>
#include <openssl/ssl.h>

pthread_mutex_t lock;

void *libctx_init(void *arg)
{
  int           rv;
  int           i;
  int           num_loops = (int)arg;
  OSSL_LIB_CTX *libctx = NULL;
  const char   *path = "/ade/mmiyashi_network_openssl/oracle/crypto/lib/openssl.cnf";

  for (i= 0; i < num_loops; i++)
  {
    printf("[%u][%d] LoadConfig called\n", pthread_self(), i);

    libctx = OSSL_LIB_CTX_new();
    if (libctx == NULL)
    {
       printf("[%u] Failed to allocate LIB_CTX\n", pthread_self());
    }
   
    //pthread_mutex_lock(&lock);
    rv = OSSL_LIB_CTX_load_config(libctx, path);
    //pthread_mutex_unlock(&lock);
    if (rv != 1)
    {
       printf("[%u] Failed to load config\n", pthread_self());
    }

    printf("[%u][%d] LoadConfig successful\n", pthread_self(), i);
  }
}

int main(int argc, char *argv[])
{
  int        rv;
  int        i;
  int        num_loops = 1;
  int        num_threads = 5;
  pthread_t  threads[1024];

  if (argc >= 2)
  {
     num_threads = atoi(argv[1]);
     if (num_threads > 1024)
     {
         printf("num_threads[%d] too largs. Reset to 1024\n", num_threads);
         num_threads = 1024;
     }
  }
  if (argc >= 3)
  {
     num_loops = atoi(argv[2]);
  }
  if (pthread_mutex_init(&lock, NULL) != 0) {
    printf("mutex init has failed\n");
    return 1;
  }

  SSL_load_error_strings();
  SSL_library_init();

  fprintf(stderr, "Initialize libctx\n");
  for (i = 0; i < num_threads; i++)
  {
    rv = pthread_create(&threads[i], NULL, libctx_init, (void *)num_loops);
    if (rv != 0) {
      printf("[%d]: pthread_create filed\n", i);
    }
  }

  for (i = 0; i < num_threads; i++)
  {
    pthread_join(threads[i], NULL);
  }

  return 0;
}
```

Is this an multi-threaded issue in OpenSSL? or an issue in the test code.

The issue is seen with OpenSSL 3.0.0.

Thank you,

-- misaki

### PR DESCRIPTION
**Pull request by hlandau**
Fixes #18226.

This is a start, but  I have some questions remaining about it.

Firstly, this adds new APIs to allow libctx to be specified, but that punts this to 3.1, not 3.0. Which raises the question of how (or if) we want to fix this in 3.0.

Secondly, it's worth noting that existing code &mdash;  including `OPENSSL_load_builtin_modules` &mdash; will only load into the default libctx. Do we want to change all of our own implementations to fix this? It means adding `OPENSSL_load_builtin_modules_ex` as well, I guess.

### CODE REVIEW

File: crypto/context.c [hunk 1]
```diff
@@ -43,6 +43,9 @@ struct ossl_lib_ctx_st {
     void *fips_prov;
 #endif
 
+    OPENSSL_STACK *supported_conf_modules;
+    OPENSSL_STACK *initialized_conf_modules;
```
Comment (paulidale) [inline crypto/context.c:47]: These should be `STACK_OF(type)`.
`OPENSSL_STACK` circumvents type checking.

File: crypto/context.c [hunk 2]
```diff
@@ -89,6 +92,15 @@ static int context_init(OSSL_LIB_CTX *ctx)
         goto err;
     exdata_done = 1;
 
+    /* Initialize module lists for CONF. */
+    ctx->supported_conf_modules = OPENSSL_sk_new_null();
```
Comment (paulidale) [inline crypto/context.c:96]: Likewise, use the safe stack wrappers here and below.

File: doc/man3/CONF_modules_free.pod [hunk 3]
```diff
@@ -25,12 +28,16 @@ configuration modules.  Normally, in versions of OpenSSL prior to 1.1.0,
 applications called
 CONF_modules_free() at exit to tidy up any configuration performed.
 
-CONF_modules_finish() calls each configuration modules B<finish> handler
-to free up any configuration that module may have performed.
+CONF_modules_finish_ex() calls each configuration modules B<finish> handler to
```
Comment (paulidale) [inline doc/man3/CONF_modules_free.pod:31]: _module's_

File: doc/man3/CONF_modules_free.pod [hunk 4]
```diff
@@ -25,12 +28,16 @@ configuration modules.  Normally, in versions of OpenSSL prior to 1.1.0,
...
+free up any configuration that module may have performed. CONF_modules_finish()
+is equivalent to calling CONF_modules_finish_ex() with the default library
```
Comment (paulidale) [inline doc/man3/CONF_modules_free.pod:33]: with or on (the default library context)?

File: doc/man3/CONF_modules_free.pod [hunk 5]
```diff
@@ -25,12 +28,16 @@ configuration modules.  Normally, in versions of OpenSSL prior to 1.1.0,
...
+context.
 
-CONF_modules_unload() finishes and unloads configuration modules. If
-B<all> is set to B<0> only modules loaded from DSOs will be unloads. If
-B<all> is B<1> all modules, including built-in modules will be unloaded.
+CONF_modules_unload_ex() finishes and unloads configuration modules. If B<all>
+is set to B<0> only modules loaded from DSOs will be unloads. If B<all> is B<1>
```
Comment (paulidale) [inline doc/man3/CONF_modules_free.pod:37]: s/unloads/unloaded/

File: doc/man3/CONF_modules_load_file.pod [hunk 6]
```diff
@@ -39,9 +45,18 @@ can be overridden by B<config_diagnostics> as described in L<config(5)>.
...
+CONF_modules_load_ex() is identical to CONF_modules_load_file() except it
 reads configuration information from B<cnf>.
 
+CONF_modules_load() is identical to CONF_modules_load_ex() except it
+has a NULL library context.
+
+CONF_module_add_ex() registers a module into a library context using the given
+name, initialization and finish functions.
+
+CONF_module_add() is identical to CONF_modules_add_Ex() except it
```
Comment (paulidale) [inline doc/man3/CONF_modules_load_file.pod:57]: CONF_modules_add_**e**x

**Comment by hlandau** [comment]
This is likely to be superseded, but updated in any case.

**Comment by t8m** [comment]
Can we close this?

**Comment by hlandau** [comment]
Yes.

**Pull request closed without merging**